File: javascript/say/say.js
export const say = (n) => {
  // Check if number is within the valid range
  if (n < 0 || n > 999999999999) {
    throw new Error('Number must be between 0 and 999,999,999,999.');
  }

  // Handle zero as a special case
  if (n === 0) {
    return 'zero';
  }

  // Break the number into chunks of thousands
  const chunks = [];
  let remaining = n;
  
  while (remaining > 0) {
    chunks.push(remaining % 1000);
    remaining = Math.floor(remaining / 1000);
  }

  // Scale words for each chunk position
  const scales = ['', 'thousand', 'million', 'billion'];
  
  // Process each chunk and add the appropriate scale
  const parts = [];
  
  for (let i = 0; i < chunks.length; i++) {
    if (chunks[i] !== 0) {
      const chunkText = convertLessThanThousand(chunks[i]);
      const scale = scales[i];
      
      if (scale) {
        parts.unshift(`${chunkText} ${scale}`);
      } else {
        parts.unshift(chunkText);
      }
    }
  }
  
  return parts.join(' ');
};

// Convert a number less than 1000 to words
function convertLessThanThousand(n) {
  if (n === 0) {
    return '';
  }
  
  const hundreds = Math.floor(n / 100);
  const remainder = n % 100;
  
  let result = '';
  
  if (hundreds > 0) {
    result += `${convertLessThanTwenty(hundreds)} hundred`;
    if (remainder > 0) {
      result += ' ';
    }
  }
  
  if (remainder > 0) {
    result += convertLessThanHundred(remainder);
  }
  
  return result;
}

// Convert a number less than 100 to words
function convertLessThanHundred(n) {
  if (n < 20) {
    return convertLessThanTwenty(n);
  }
  
  const tens = Math.floor(n / 10);
  const ones = n % 10;
  
  if (ones === 0) {
    return convertTens(tens);
  } else {
    return `${convertTens(tens)}-${convertLessThanTwenty(ones)}`;
  }
}

// Convert a number less than 20 to words
function convertLessThanTwenty(n) {
  const words = [
    '', 'one', 'two', 'three', 'four', 'five', 'six', 'seven', 'eight', 'nine', 'ten',
    'eleven', 'twelve', 'thirteen', 'fourteen', 'fifteen', 'sixteen', 'seventeen', 'eighteen', 'nineteen'
  ];
  
  return words[n];
}

// Convert tens (20, 30, 40, etc.) to words
function convertTens(n) {
  const words = [
    '', '', 'twenty', 'thirty', 'forty', 'fifty', 'sixty', 'seventy', 'eighty', 'ninety'
  ];
  
  return words[n];
}

File: go/say/say.go
package say

import (
	"strings"
)

// Say converts an integer to its English representation
func Say(n int64) (string, bool) {
	// Check if the number is within the valid range
	if n < 0 || n >= 1000000000000 {
		return "", false
	}

	// Special case for zero
	if n == 0 {
		return "zero", true
	}

	// Convert the number to English
	return sayNumber(n), true
}

// sayNumber converts a positive number to its English representation
func sayNumber(n int64) string {
	// Arrays for special cases
	ones := []string{"", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine",
		"ten", "eleven", "twelve", "thirteen", "fourteen", "fifteen", "sixteen", "seventeen", "eighteen", "nineteen"}
	tens := []string{"", "", "twenty", "thirty", "forty", "fifty", "sixty", "seventy", "eighty", "ninety"}
	scales := []string{"", "thousand", "million", "billion"}

	// Handle numbers less than 100
	if n < 20 {
		return ones[n]
	}
	if n < 100 {
		if n%10 == 0 {
			return tens[n/10]
		}
		return tens[n/10] + "-" + ones[n%10]
	}

	// Handle numbers less than 1000
	if n < 1000 {
		if n%100 == 0 {
			return ones[n/100] + " hundred"
		}
		return ones[n/100] + " hundred " + sayNumber(n%100)
	}

	// Break the number into chunks of thousands
	var result strings.Builder
	var chunks []int64

	// Extract chunks of up to 3 digits
	temp := n
	for temp > 0 {
		chunks = append(chunks, temp%1000)
		temp /= 1000
	}

	// Build the result string from the chunks
	for i := len(chunks) - 1; i >= 0; i-- {
		if chunks[i] > 0 {
			if result.Len() > 0 {
				result.WriteString(" ")
			}
			result.WriteString(sayNumber(chunks[i]))
			if i > 0 { // Add scale word if not the last chunk
				result.WriteString(" " + scales[i])
			}
		}
	}

	return result.String()
}
